File: 10-heap/00621-Task_Scheduler/00621-task_scheduler.go
// 621: Task Scheduler
// https://leetcode.com/problems/task-scheduler

package main

import "fmt"

// SOLUTION
func leastInterval(tasks []byte, n int) int {
	if n == 0 {
		return len(tasks)
	}

	counts := make([]int, 26)
	for _, task := range tasks {
		counts[task-'A']++
	}

	var max_count, time int
	for _, c := range counts {
		if c > max_count {
			max_count = c
			time = 1
		} else if c == max_count {
			time++
		}
	}

	result := (n+1)*(max_count-1) + time

	if result > len(tasks) {
		return result
	} else {
		return len(tasks)
	}
}

func main() {
	// INPUT
	tasks := []byte{'A', 'C', 'A', 'B', 'D', 'B'}
	n := 1

	// OUTPUT
	result := leastInterval(tasks, n)
	fmt.Println(result)
}

File: 10-heap/00621-Task_Scheduler/00621-task_scheduler.py
# 621: Task Scheduler
# https://leetcode.com/problems/task-scheduler


class Solution:
    # SOLUTION
    def leastInterval(self, tasks: list[str], n: int) -> int:
        counts = [0] * 26
        for task in tasks:
            counts[ord(task) - ord('A')] += 1
        counts.sort()

        chunk = counts[25] - 1
        time = chunk * n

        for i in range(24, -1, -1):
            time -= min(chunk, counts[i])

        return len(tasks) + time if time >= 0 else len(tasks)


if __name__ == "__main__":
    o = Solution()

    # INPUT
    tasks = ['A','C','A','B','D','B']
    n = 1

    # OUTPUT
    result = o.leastInterval(tasks, n)
    print(result)
